dll/cpl/sysdm/environment.c

```diff
@@ -144,16 +144,13 @@ GatherDataFromListView(HWND hwndListView,
         dwValueLength += _tcslen(szData);
     }
 
-    /* Reallocate the value buffer, regrowing it if necessary */
-    if (!VarData->lpRawValue || (_tcslen(VarData->lpRawValue) < dwValueLength))
-    {
-        if (VarData->lpRawValue)
-            GlobalFree(VarData->lpRawValue);
+    /* Reallocate the value buffer */
+    if (VarData->lpRawValue)
+        GlobalFree(VarData->lpRawValue);
 
-        VarData->lpRawValue = GlobalAlloc(GPTR, (dwValueLength + 1) * sizeof(TCHAR));
-        if (!VarData->lpRawValue)
-            return 0;
-    }
+    VarData->lpRawValue = GlobalAlloc(GPTR, (dwValueLength + 1) * sizeof(TCHAR));
+    if (!VarData->lpRawValue)
+        return 0;
 
```

Comment:
Simpler fix than all these changes:
1. Keep original code,
2. Add the actual missing `VarData->lpRawValue[0] = _T('\0');` just after that block, and before the copy.

The actual problem was, that since you are using the concatenation functions below, the code was trying to concatenate onto the existing data.

Reply:
> Simpler fix than all these changes:
> 
> 1. Keep original code,
> 2. Add the actual missing `VarData->lpRawValue[0] = _T('\0');` just after that block, and before the copy.
> 
> The actual problem was, that since you are using the concatenation functions below, the code was trying to concatenate onto the existing data.

@HBelusca The reason why I didn't do this since I thought that since we aren't required the old values at all and going for the new values, it would be much simpler to just reallocating it again. (since even if it is bigger or smaller, it would get the desired size)
Nonetheless I am taking your suggestion.